internal/provider/cached_image_resource.go

```diff
@@ -293,128 +295,246 @@ func (r *CachedImageResource) Configure(ctx context.Context, req resource.Config
+		switch key {
+
+		// These options may not be overridden.
+		case "ENVBUILDER_CACHE_REPO", "ENVBUILDER_GIT_URL":
+			diags.AddAttributeWarning(path.Root("extra_env"),
+				"Cannot override required environment variable",
+				fmt.Sprintf("The key %q in extra_env cannot be overridden.", key),
+			)
+			continue
+
+		default:
+			// Check if the option was set on the provider data model and generate a warning if so.
+			if _, overridden := overrides[key]; overridden {
+				diags.AddAttributeWarning(path.Root("extra_env"),
+					"Overriding provider environment variable",
+					fmt.Sprintf("The key %q in extra_env overrides an option set on the provider.", key),
+				)
+			}
+
+			// XXX: workaround for serpent behaviour where calling Set() on a
+			// string slice will append instead of replace: set to empty first.
+			if key == "ENVBUILDER_IGNORE_PATHS" {
+				_ = optsMap[key].Set("")
+			}
+
+			opt, found := optsMap[key]
+			if !found {
+				// ignore unknown keys
+				continue
+			}
+
+			if err := opt.Set(val); err != nil {
+				diags.AddAttributeError(path.Root("extra_env"),
+					"Invalid value for environment variable",
+					fmt.Sprintf("The key %q in extra_env has an invalid value: %s", key, err),
 				)
-			default:
-				if _, ok := env[key]; ok {
-					// This is a warning because it's possible that the user wants to override
-					// a value set by the provider.
-					diag.AddAttributeWarning(path.Root("extra_env"),
-						"Overriding provider environment variable",
-						fmt.Sprintf("The key %q in extra_env overrides an environment variable set by the provider.", key),
-					)
-				}
-				env[key] = tfValueToString(elem)
 			}
 		}
 	}
+	return diags
+}
+
+func computeEnvFromOptions(opts eboptions.Options, extraEnv map[string]string) map[string]string {
+	allEnvKeys := make(map[string]struct{})
+	for _, opt := range opts.CLI() {
+		if opt.Env == "" {
+			continue
+		}
+		allEnvKeys[opt.Env] = struct{}{}
+	}
+
+	// Only set the environment variables from opts that are not legacy options.
+	// Legacy options are those that are not prefixed with ENVBUILDER_.
+	// While we can detect when a legacy option is set, overriding it becomes
+	// problematic. Erring on the side of caution, we will not override legacy options.
+	isEnvbuilderOption := func(key string) bool {
+		switch key {
+		case "CODER_AGENT_URL", "CODER_AGENT_TOKEN", "CODER_AGENT_SUBSYSTEM":
+			return true // kinda
+		default:
+			return strings.HasPrefix(key, "ENVBUILDER_")
+		}
+	}
+
+	computed := make(map[string]string)
+	for _, opt := range opts.CLI() {
+		if opt.Env == "" {
+			continue
+		}
+		// TODO: remove this check once support for legacy options is removed.
+		if !isEnvbuilderOption(opt.Env) {
+			continue
+		}
+		var val string
+		if sa, ok := opt.Value.(*serpent.StringArray); ok {
+			val = strings.Join(sa.GetSlice(), ",")
+		} else {
+			val = opt.Value.String()
+		}
+
+		switch val {
+		case "", "false", "0":
+			// Skip zero values.
+			continue
+		}
```

Comment:
AFAIK all of our boolean-ish options are `false` by default.

Reply:
This is a bit flimsy IMHO; could we be more explicit about the types we want to skip, or use reflect to determine if a zero value is set?

Reply:
Reflect would definitely be an option here; I chose to just hard-code these for simplicity.